### PR TITLE
Add support for integer, double, boolean, map/pojo, and array type in…

### PR DESCRIPTION
**Pull request by ChrisFlanTheMan**
…put variables when correlating events

### CODE REVIEW

**Review by LeoZ100 — APPROVED** [review]

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private Object evaluatePOJONode(POJONode pojoNode) {
+        try {
+            String jsonString = objectMapper.writeValueAsString(pojoNode.getPojo());
+            Map<String, Object> objectMap = objectMapper.readValue(jsonString, Map.class);
+            return objectMap;
+        } catch (JsonProcessingException ex) {
+            throw new RuntimeException(String.format("Unable to parse POJONode value %s into Map, the following exception occurred: %s",
+                pojoNode.getPojo().toString(), ex.getMessage()));
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java:59]: Don't call `ex.getMessage()`, instead pass the json processing exception as the cause of your runtime exception

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java [hunk 2]
```diff
@@ -39,25 +39,26 @@ public static CorrelationData buildCorrelationData(GenericMessage genericMessage
...
     }
 
-    private static String evaluateExpression(DocumentContext documentContext, String expression) {
+    private static Object evaluateExpression(DocumentContext documentContext, String expression) {
         // Since we are using JacksonJsonNodeJsonProvider we need to convert
         // the result of the JsonPath into the value we need
         JsonNode node = documentContext.read(expression);
-        return node.textValue();
+
+        return new JsonNodeEvaluator().evaluateNode(node);
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java:62]: I don't think you have to new this object up each time

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        } else if (node.isBoolean()) {
+            return node.booleanValue();
+        } else if (node.isDouble()) {
+            return node.doubleValue();
+        } else {
+            return node.toString();
+        }
+    }
+
+    private Object evaluateArrayNode(ArrayNode arrayNode) {
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java:40]: you can simplify this method by using `arrayNode.forEach(node -> ...)`

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+public final class JsonNodeEvaluator {
+
+    private ObjectMapper objectMapper = new ObjectMapper()
+        .configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true);
+
+    public JsonNodeEvaluator() {
+    }
+
+    public Object evaluateNode(JsonNode node) {
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java:22]: It seems you might be missing a check for `null` here

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java (hunk 2 above)
Comment (dare2try) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java:62]: If `null` is a possibility I would check here that way you don't throw a `NRE`

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java [hunk 5]
```diff
@@ -39,25 +39,26 @@ public static CorrelationData buildCorrelationData(GenericMessage genericMessage
         Map<String, String> matchVariables = new HashMap<>();
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java:39]: Should `matchVariables` also be `Map<String, Object>`?

**Review by dare2try — CHANGES_REQUESTED** [review]

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java (hunk 5 above)
Comment (ChrisFlanTheMan) [reply messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/CorrelationDataUtils.java:39]: Yep, i can extend that map as well

File: messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java [hunk 6]
```diff
@@ -7,19 +7,22 @@
...
 import com.fasterxml.jackson.databind.node.POJONode;
 import java.util.ArrayList;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 
 public final class JsonNodeEvaluator {
 
-    private ObjectMapper objectMapper = new ObjectMapper()
+    private static final ObjectMapper objectMapper = new ObjectMapper()
```
Comment (dare2try) [inline messaging/src/main/java/com/ultimatesoftware/workflow/messaging/correlation/JsonNodeEvaluator.java:15]: I would not make this `static`

**Review by dare2try — APPROVED** [review]
Just one small comment about the use of static but otherwise it's good

**Pull request merged**